experimental/uefi/channel/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#![no_std]
+#![feature(never_type)]
+
+extern crate alloc;
+
+use alloc::{vec, vec::Vec};
+use anyhow::{bail, Context};
+use ciborium_io::{Read, Write};
+use oak_remote_attestation_sessions::{SessionId, SESSION_ID_LENGTH};
+
+/// Length of the entire frame, including the header.
+pub type FrameLength = u32;
+const FRAME_BYTES_SIZE: usize = 4;
+
+pub struct Frame {
+    pub body: Vec<u8>,
+}
+
+pub struct Framed<T: Read + Write> {
+    inner: T,
+}
+
+impl<T> Framed<T>
+where
+    T: Read<Error = anyhow::Error> + Write<Error = anyhow::Error>,
+{
+    pub fn new(channel: T) -> Self {
+        Self { inner: channel }
+    }
+
+    pub fn read_frame(&mut self) -> anyhow::Result<Frame> {
+        let length = {
+            let mut length_bytes = [0; FRAME_BYTES_SIZE];
+            self.inner.read_exact(&mut length_bytes)?;
+            FrameLength::from_le_bytes(length_bytes)
+        };
+
+        let body = {
+            let body_length: u32 = length - FRAME_BYTES_SIZE as u32;
+            let mut body: Vec<u8> = vec![
+                0;
+                usize::try_from(body_length).expect(
+                    "the supported pointer size is smaller than the frame length"
+                )
+            ];
+            self.inner.read_exact(&mut body)?;
+            body
+        };
+
+        Ok(Frame { body })
+    }
+
+    pub fn write_frame(&mut self, frame: Frame) -> anyhow::Result<()> {
+        let length_bytes = {
+            let length = FrameLength::try_from(FRAME_BYTES_SIZE + frame.body.len())
+                .map_err(anyhow::Error::msg)
+                .context("the frame is too large")?;
+            length.to_le_bytes()
+        };
+        self.inner.write_all(&length_bytes)?;
+        self.inner.write_all(&frame.body)?;
+        Ok(())
+    }
+}
+
+pub struct SerializeableRequest {
```

Comment:
I don't think this should be moved here, as it means that the channel crate depends on remote attestation and crypto libraries. I think it is fine to leave it in the runtime for now, and later when we add more serialisable objects (or add full-fledged IDL support) move this elsewhere.

A side note: I don't think the definition of the session ID should be in a crate that forces a dependency on crypto libraries, so as part of the future move to an IDL-related crate the dependency should probably be reversed (i.e. SessionId should be defined in the new crate and remote_attestation_session should depend on it)

Reply:
Thinking about it a bit more, moving this struct to the `remote_attestation_sessions` crate might be useful as an initial step, seeing that this is tightly coupled with attestation sessions. It would hopefully remove the need for the loader to depend on the runtime. The runtime would still depend on the crypto libraries (which it should not need) for now, but that can be resolved later when moving this to an IDL-related crate.

Reply:
Clarity of the mental model, which you mentioned is probalby the most important, especially to simplify the review process for external reviewers.

Clearer separation of dependencies and reposnsibilities is usually a desirable property. if the loader/launcher (running outside the enclave) depends on the runtime (running inside the enclave) it would need to be rebuilt even if an unrelated part of the runtime changed. It also inherits a bunch of other dependencies, which are built but unused. Having more indirect dependencies also complicates the review process, as the reviewer needs to figure out which indirect dependencies are actually used.

We also want to start adding claims (in a human readable format for reviewers) to explain which properties are enforced/ensured by each crate, how this contributes to the overall security/privacy, story and how this is achieved in the code. This becomes much more difficult if there are long chains of indirect depenencies or complex crates doing multiple different things. Every time a dependency changes these need to be reviewed to ensure they are still valid.

No_std support could also be affected by some indirect dependencies, e.g. if the dependency chain was the other way around some std-related features enabled for the loader could stop the runtime from building for a bare-metal target. Porting the runtime to no_std is significantly simplified by the earlier work of splitting components into smaller crates with fewer dependencies.